win32/tests/file_win32_ut/file_win32_ut.c

```diff
@@ -0,0 +1,952 @@
+        .SetReturn(h_event);
+    STRICT_EXPECTED_CALL(mock_StartThreadpoolIo(IGNORED_ARG))
+        .CaptureArgumentValue_pio(&captured_ptp_io);
+    STRICT_EXPECTED_CALL(mock_ReadFile(IGNORED_ARG, destination, size, NULL, IGNORED_ARG))
+        .SetReturn(FALSE);
+    STRICT_EXPECTED_CALL(mock_GetLastError())
+        .SetReturn(ERROR_IO_PENDING);
+
+    ///act
+    FILE_READ_ASYNC_RESULT result = file_read_async(file_handle, destination, size, position, user_callback, NULL);
+
+    ///assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_ARE_EQUAL(FILE_READ_ASYNC_RESULT, FILE_READ_ASYNC_OK, result);
+    
+    ///cleanup
+    file_destroy(file_handle);
+}
+
+/*Tests_SRS_FILE_43_036: [ If the call to read the file fails, file_read_async shall fail and return FILE_READ_ASYNC_READ_ERROR. ]*/
+/*Tests_SRS_FILE_WIN32_43_025: [ file_read_async shall call StartThreadpoolIo.]*/
+/*Tests_SRS_FILE_WIN32_43_055: [ file_read_async shall create an event by calling CreateEvent.]*/
+/*Tests_SRS_FILE_WIN32_43_028: [ file_read_async shall allocate an OVERLAPPED struct and populate it with the created event and position.]*/
+/*Tests_SRS_FILE_WIN32_43_026: [ file_read_async shall allocate a context to store the allocated OVERLAPPED struct, destination, handle, size, user_callback and user_context ]*/
+/*Tests_SRS_FILE_WIN32_43_029: [ file_read_async shall call ReadFile with handle, destination, size and the allocated OVERLAPPED struct.]*/
+/*Tests_SRS_FILE_WIN32_43_031: [ If ReadFile fails synchronously and GetLastError does not indicate ERROR_IO_PENDING then file_read_async shall fail, call CancelThreadpoolIo and return FILE_READ_ASYNC_READ_ERROR. ]*/
+TEST_FUNCTION(file_read_async_fails_synchronously)
+{
+    ///arrange
+    PTP_WIN32_IO_CALLBACK captured_callback;
+    FILE_HANDLE file_handle = get_file_handle("file_read_async_fails_synchronously.txt", &captured_callback);
+    unsigned char destination[10];
+    FILE_CB user_callback = (FILE_CB)43;
+    uint32_t size = 10;
+    uint64_t position = 5;
+    HANDLE h_event = (HANDLE)44;
+    void* io;
+    PTP_IO captured_ptp_io;
+
+    STRICT_EXPECTED_CALL(malloc(IGNORED_ARG))
+        .CaptureReturn(&io);
+    STRICT_EXPECTED_CALL(mock_CreateEvent(IGNORED_ARG, IGNORED_ARG, IGNORED_ARG, IGNORED_ARG))
+        .SetReturn(h_event);
+    STRICT_EXPECTED_CALL(mock_StartThreadpoolIo(IGNORED_ARG))
+        .CaptureArgumentValue_pio(&captured_ptp_io);
+    STRICT_EXPECTED_CALL(mock_ReadFile(IGNORED_ARG, destination, size, NULL, IGNORED_ARG))
+        .SetReturn(FALSE);
+    STRICT_EXPECTED_CALL(mock_GetLastError())
+        .SetReturn(ERROR_IO_INCOMPLETE);
+    STRICT_EXPECTED_CALL(mock_CancelThreadpoolIo(IGNORED_ARG))
+        .ValidateArgumentValue_pio(&captured_ptp_io);
+    STRICT_EXPECTED_CALL(mock_CloseHandle(h_event));
+    STRICT_EXPECTED_CALL(free(IGNORED_ARG))
+        .ValidateArgumentValue_ptr(&io);

```

Comment:
Could probably put this in a helper (and have some parameters to generalize it across all the calls).

Reply:
I decided to keep it as is because the calls are different depending on whether the [Read/Write]File returns synchronously or asynchronously

---
In reply to: [456721090](https://github.com/Azure/azure-c-pal/pull/10#discussion_r456721090) [](ancestors = 456721090)